client/app/components/text-area.js

```diff
@@ -0,0 +1,9 @@
+import Component from '@glimmer/component';
+
+export default class TextAreaComponent extends Component {
+  rows = this.args.rows || '5';
```

Comment:
I didn't want to reach for the getter here because this style here does the trick in terms of setting default values on initial component render. It's similar to the recommended way of setting default values for internal variables.

Using a getter here means that we expect the incoming argument `this.args.rows` to be dynamic and change after first render -- this is my understanding from working with component getters and args so far. 

I didn't think of this when initially writing this class, but we can consider how arguments like `rows`, `maxLength`, and even `model` or `property` are very unlikely to change after first render. I might be stretching this argument, but it might even be bad form design to allow rows or maxLength arguments to change after first render. We could even say that this style disallows those arguments to be dynamic.

Reply:
Thinking more on default values, I think you make a really interesting point here: 

> It's similar to the recommended way of setting default values for internal variables.

But there's an important distinction between "internal" (local? private?) variables and arguments: the latter are totally immutable. If you try to say `this.args.lol = "no"`, it will throw an exception. Now, that's fine to know that and not do that in practice, but a getter will actually reinforce the immutability of arguments:

```
get rows() {
  return this.args.rows || '5';
}
```

There's no way to _set_ this property because we've explicitly said it's a getter and did not specify a way to _set_ it.

As for the likeliness of things needing to re-render, won't the model and property itself very likely change after a first render? When people are typing stuff? Preferring a non-reactive render behavior would actually be more surprising to a future maintainer. The entire point of using a framework is having reactivity "just work". If we want to make an exception to that, I think we need a stronger reason. 

As for `rows` — which is a native HTML text area attribute —  you may be correct in predicting that it's unlikely to ever change throughout the session, but why not just preserve the default behavior of native HTML elements (as these are native attributes of the text area element)? HTML is itself reactive — if I update an attribute in the DOM, it will visually update that element. Having an attribute for our component which shares the same name as the native element attr and _doesn't_ update DOM is more confusing to me.

In fact, I'm still not sure we should even be keeping these default arguments. With `...attributes`, we can pass in those values from the consuming end without needing to deal with defaults. And if you want to make a default `rows` value, you could say this:

```
rows="5"
```

If someone wants to override that, then they can just pass that attribute in from the consuming end. That way, we don't introduce any new API surface that a maintainer has to learn. It's "Just HTML".

I think my main concern here is having this component class file at all. I don't think we need it yet :)

Reply:
> Thinking more on default values, I think you make a really interesting point here:
> 
> > It's similar to the recommended way of setting default values for internal variables.
> 
> But there's an important distinction between "internal" (local? private?) variables and arguments: the latter are totally immutable. If you try to say `this.args.lol = "no"`, it will throw an exception. Now, that's fine to know that and not do that in practice, but a getter will actually reinforce the immutability of arguments:
> 
> ```
> get rows() {
>   return this.args.rows || '5';
> }
> ```
> 
> There's no way to _set_ this property because we've explicitly said it's a getter and did not specify a way to _set_ it.
> 

This is a really good point. I didn't think of that but you're right -- we shouldn't "undo" the immutability of the args by reassigning it to a non-getter. 

> As for the likeliness of things needing to re-render, won't the model and property itself very likely change after a first render? When people are typing stuff? Preferring a non-reactive render behavior would actually be more surprising to a future maintainer. The entire point of using a framework is having reactivity "just work". If we want to make an exception to that, I think we need a stronger reason.
> 
> As for `rows` — which is a native HTML text area attribute — you may be correct in predicting that it's unlikely to ever change throughout the session, but why not just preserve the default behavior of native HTML elements (as these are native attributes of the text area element)? HTML is itself reactive — if I update an attribute in the DOM, it will visually update that element. Having an attribute for our component which shares the same name as the native element attr and _doesn't_ update DOM is more confusing to me.
> 

For form inputs, I don't think the model and property will change. But from a developer perspective (not a UI/design perspective), I do see how it will be nice for the components to maintain their reactivity in all cases.... just in case the developer needs to codes for an edge case in UI/design. 

> In fact, I'm still not sure we should even be keeping these default arguments. With `...attributes`, we can pass in those values from the consuming end without needing to deal with defaults. And if you want to make a default `rows` value, you could say this:
> 
> ```
> rows="5"
> ```
> 

You're totally right about how `rows` should be handled by splattributes! 
I do like defaults for `maxlength` and `showCounter`. Without the defaults, how do we require consumers to pass in `maxlength` and `showCounter` so they aren't undefined? 

> If someone wants to override that, then they can just pass that attribute in from the consuming end. That way, we don't introduce any new API surface that a maintainer has to learn. It's "Just HTML".
> 
> I think my main concern here is having this component class file at all. I don't think we need it yet :)